store.go

```diff
@@ -996,33 +996,35 @@ func (s *store) load() error {
 
 	s.containerStore = rcs
 
-	additionalImageStores := s.graphDriver.AdditionalImageStores()
+	additionalImageStores, additionalGraphDrivers := s.graphDriver.AdditionalImageStores()
 	if s.imageStoreDir != "" {
 		additionalImageStores = append([]string{s.graphRoot}, additionalImageStores...)
 	}
 
 	for _, store := range additionalImageStores {
-		gipath := filepath.Join(store, driverPrefix+"images")
-		var ris roImageStore
-		// both the graphdriver and the imagestore must be used read-write.
-		if store == s.imageStoreDir || store == s.graphRoot {
-			imageStore, err := newImageStore(gipath)
-			if err != nil {
-				return err
-			}
-			s.rwImageStores = append(s.rwImageStores, imageStore)
-			ris = imageStore
-		} else {
-			ris, err = newROImageStore(gipath)
-			if err != nil {
-				if errors.Is(err, syscall.EROFS) {
-					logrus.Debugf("Ignoring creation of lockfiles on read-only file systems %q, %v", gipath, err)
-					continue
+		for _, driver := range additionalGraphDrivers {
```

Comment:
This is trying a cartesian product of (store, driver) _and all combinations must exist_, otherwise this fails. 

I can’t see how that makes sense even with one additional store, let alone multiple. Am I missing something?

I would have expected `.AdditionalImageStores()` to return a `[]struct InternalAdditionalStoreSpec { storePath String, driver stringOrMaybeInterface }` or something like that. (Well, modulo the pre-existing thing that the division of responsibility for parsing/processing these options between `store.go` and the graph driver seems convoluted to me.)

Reply:
`newImageStore` automatically creates the directory.  This is a pre-existing behavior that I've kept in the current version.

I agree it is not ideal, I will add a patch to just ignore stores that do not exist instead of silently creating them

Reply:
That means silent failures on incorrect paths in the config. (We do validate their existence.)

The previous behavior of automatically creating a directory seems undesirable to me; but if we find _no_ matching driver, shouldn’t we fail outright instead of silently ignoring the store?

Also the logic in `load()` and `getROLayerStoresLocked()` should be consistent; that’s also easier if the graph driver detection is moved behind `driver.AdditionalImageStores` (I suppose into overlay’s `parseOptions` or even earlier).